packages/cli/lib/base-generator.js

```diff
@@ -28,11 +31,218 @@ module.exports = class BaseGenerator extends Generator {
+          if (line === 'EOF') {
+            rl.close();
+          } else {
+            lines.push(line);
+          }
+        })
+        .on('close', () => {
+          if (err) return;
+          const jsonStr = lines.join('\n');
+          try {
+            const json = JSON.parse(jsonStr);
+            resolve(json);
+          } catch (e) {
+            if (!process.stdin.isTTY) {
+              debug(e, jsonStr);
+            }
+            reject(e);
+          }
+        })
+        .on('error', e => {
+          err = e;
+          rl.close();
+          reject(err);
+        });
+    });
+  }
+
+  async setOptions() {
+    let opts = {};
+    const jsonFile = this.options.config;
+    try {
+      if (jsonFile === 'stdin' || !process.stdin.isTTY) {
+        this.options['skip-optional-prompts'] = true;
+        opts = await this._readJSONFromStdin();
+      } else if (typeof jsonFile === 'string') {
+        opts = this.fs.readJSON(path.resolve(process.cwd(), jsonFile));
+      }
+    } catch (e) {
+      this.exit(e);
+      return;
+    }
+    if (!(typeof opts === 'object')) {
+      this.exit('Invalid config file: ' + jsonFile);
+      return;
+    }
+    for (const o in opts) {
+      if (this.options[o] == null) {
+        this.options[o] = opts[o];
+      }
+    }
+  }
+
+  /**
+   * Override the base prompt to skip prompts with default answers
+   * @param questions One or more questions
+   */
+  async prompt(questions) {
+    // Normalize the questions to be an array
+    if (!Array.isArray(questions)) {
+      questions = [questions];
+    }
+    if (!this.options['skip-optional-prompts']) {
+      if (!process.stdin.isTTY) {
+        this.log(
+          chalk.red('The stdin is not a terminal. No prompt is allowed.'),
+        );
+        process.exit(1);
+      }
+      // Non-express mode, continue to prompt
+      return await super.prompt(questions);
+    }
+
+    const answers = Object.assign({}, this.options);
+
+    // Check if a question can be skipped in `express` mode
+    const canBeSkipped = q =>
+      q.default != null || // Having a default value
+      this.options[q.name] != null || // Configured in options
+      q.type === 'list' || // A list
+      q.type === 'rawList' || // A raw list
+      q.type === 'checkbox' || // A checkbox
+      q.type === 'confirm'; // A confirmation
+
+    // Get the default answer for a question
+    const defaultAnswer = async q => {
+      let def = q.default;
+      if (typeof q.default === 'function') {
+        def = await q.default(answers);
+      }
+      let defaultVal = def;
+
+      if (def == null) {
+        // No `default` is set for the question, check existing answers
+        defaultVal = answers[q.name];
+        if (defaultVal != null) return defaultVal;
+      }
+
+      if (q.type === 'confirm') {
+        return defaultVal != null ? defaultVal : true;
+      }
+      if (q.type === 'list' || q.type === 'rawList') {
+        // Default to 1st item
```

Comment:
Handling of `list`/`rawList` is a prime candidate to be extracted into a new function.